mono/metadata/threads.c

```diff
@@ -1670,19 +1680,45 @@ mono_thread_set_name_internal (MonoInternalThread *this_obj, MonoString *name, g
 	UNLOCK_THREAD (this_obj);
 
 	if (this_obj->name && tid) {
-		char *tname = mono_string_to_utf8_checked (name, error);
-		return_if_nok (error);
+		const char *tname = 0;
+		if (utf8)
+			tname = utf8;
+		else {
+			tname = mono_utf16_to_utf8 (utf16, length, error);
+			return_if_nok (error);
+		}
 		MONO_PROFILER_RAISE (thread_name, ((uintptr_t)tid, tname));
 		mono_native_thread_set_name (tid, tname);
-		mono_free (tname);
+		if (!utf8)
+			mono_free (MONO_CONST_CAST(char*)(tname));
 	}
 }
 
+void
+mono_thread_set_name_handle_internal (MonoInternalThread *this_obj, MonoStringHandle name, gboolean permanent, gboolean reset, MonoError *error)
+{
+	uint32_t gchandle = 0;
+	gunichar2 *chars = 0;
+	int length = 0;
+
+	if (!MONO_HANDLE_IS_NULL(name)) {
+		chars = mono_string_handle_pin_chars (name, &gchandle);
+		length = mono_string_handle_length (name);
+	}
+	
+	mono_thread_set_name_utf_internal (this_obj, length, NULL, chars, permanent, reset, error);
+
+	if (gchandle)
+		mono_gchandle_free (gchandle);
+}
+
 void 
 ves_icall_System_Threading_Thread_SetName_internal (MonoInternalThread *this_obj, MonoString *name)
```

Comment:
Better to redo the whole icall to use handles (ie, mark it with `HANDLES()` in `icall-def.h`, take the `MonoError*` argument and handles for the thread object and the name string, etc).

Reply:
I realized later this is due for more work but I was just doing the small change consistent with others for now. I'll get back to it soon enough.